src/core_functions/scalar/list/array_slice.cpp

```diff
@@ -65,108 +89,265 @@ list_entry_t SliceValue(Vector &result, list_entry_t input, int64_t begin, int64
-
-		s.ToUnifiedFormat(count, sdata);
-		b.ToUnifiedFormat(count, bdata);
-		e.ToUnifiedFormat(count, edata);
-
-		auto rdata = FlatVector::GetData<INPUT_TYPE>(result);
-		auto &rmask = FlatVector::Validity(result);
-
-		for (idx_t i = 0; i < count; ++i) {
-			auto sidx = sdata.sel->get_index(i);
-			auto bidx = bdata.sel->get_index(i);
-			auto eidx = edata.sel->get_index(i);
-
-			auto sliced = (UnifiedVectorFormat::GetData<INPUT_TYPE>(sdata))[sidx];
-			auto begin = (UnifiedVectorFormat::GetData<INDEX_TYPE>(bdata))[bidx];
-			auto end = (UnifiedVectorFormat::GetData<INDEX_TYPE>(edata))[eidx];
-
-			auto svalid = sdata.validity.RowIsValid(sidx);
-			auto bvalid = bdata.validity.RowIsValid(bidx);
-			auto evalid = edata.validity.RowIsValid(eidx);
-
-			// Try to slice
-			if (!svalid || !ClampSlice(sliced, begin, end, bvalid, evalid)) {
-				rmask.SetInvalid(i);
-			} else {
-				rdata[i] = SliceValue<INPUT_TYPE, INDEX_TYPE>(result, sliced, begin, end);
-			}
+		rdata[0] = SliceValueWithSteps<INPUT_TYPE, INDEX_TYPE>(result, sel, sliced, begin, end, step, sel_idx);
+	}
+
+	if (s && step != 0 && end - begin > 0) {
+		result_child_vector->Slice(sel, sel_length);
+	}
+}
+
+template <typename INPUT_TYPE, typename INDEX_TYPE>
+static void FindSelLength(UnifiedVectorFormat &vdata, UnifiedVectorFormat &bdata, UnifiedVectorFormat &edata,
+                          UnifiedVectorFormat &sdata, const idx_t count, idx_t &sel_length) {
+	for (idx_t i = 0; i < count; ++i) {
+		auto vidx = vdata.sel->get_index(i);
+		auto bidx = bdata.sel->get_index(i);
+		auto eidx = edata.sel->get_index(i);
+		auto sidx = sdata.sel->get_index(i);
+
+		auto sliced = ((INPUT_TYPE *)vdata.data)[vidx];
+		auto begin = ((INDEX_TYPE *)bdata.data)[bidx];
+		auto end = ((INDEX_TYPE *)edata.data)[eidx];
+		auto step = ((INDEX_TYPE *)sdata.data)[sidx];
+
+		if (step < 0) {
+			std::swap(begin, end);
+		}
+
+		if (begin == (INDEX_TYPE)NumericLimits<int64_t>::Maximum()) {
+			begin = 0;
+		}
+
+		if (end == (INDEX_TYPE)NumericLimits<int64_t>::Maximum()) {
+			end = ValueLength<INPUT_TYPE, INDEX_TYPE>(sliced);
+		}
+
+		auto step_valid = sdata.validity.RowIsValid(sidx);
+
+		auto length = 0;
+		if (step_valid && ClampSlice(sliced, begin, end)) {
+			length = CalculateSliceLength(begin, end, step, step_valid);
+		}
+		sel_length += length;
+	}
+}
+
+template <typename INPUT_TYPE, typename INDEX_TYPE>
+static void ExecuteFlatSlice(Vector &result, Vector &v, Vector &b, Vector &e, const idx_t count, optional_ptr<Vector> s,
+                             SelectionVector &sel, idx_t &sel_idx, optional_ptr<Vector> result_child_vector) {
+	UnifiedVectorFormat vdata, bdata, edata, sdata;
+	idx_t sel_length = 0;
+
+	v.ToUnifiedFormat(count, vdata);
+	b.ToUnifiedFormat(count, bdata);
+	e.ToUnifiedFormat(count, edata);
+	if (s) {
+		s->ToUnifiedFormat(count, sdata);
+		FindSelLength<INPUT_TYPE, INDEX_TYPE>(vdata, bdata, edata, sdata, count, sel_length);
+		sel.Initialize(sel_length);
+	}
```

Comment:
You can unify this and the iteration below (I might miss something). This should save a lot of code duplication and improve performance.

You can `Initialize` the selection vector with the length of the child vector (first). Then, you calculate `sel_length` while iterating. And before you slice, you first slice the selection vector like so.
```cpp
SelectionVector new_sel (sel.Slice(sel, sel_length));
result_child_vector->Slice(new_sel, sel_length);
```